### PR TITLE
MSC3006: Bot Interactions

### PR DESCRIPTION
**Pull request by MTRNord**
This MSC is part 1 of 2 on Bot interactions. There is going to be a followup on this combining reactions and the interactions of this MSC for inline links/buttons/actions :)

[Rendered](https://github.com/MTRNord/matrix-doc/blob/patch-2/proposals/3006-bot-interactions.md)

Signed-Off-By: MTRNord <info@nordgedanken.de>

Element-Web implementation: https://github.com/matrix-org/matrix-react-sdk/pull/5842

# TODOS (these came up while implementing)

- [x] State events with statekey need PL50 by default it seems (needs a change to the default PL event and probably by that a room version change)
- [x] Fallback Section was never written but should have been added

### CODE REVIEW

File: proposals/3006-bot-interactions.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
+# MSC3006: Bot Interactions
+
+Bots are currently only possible in irc style. They are entirely text based and not interactive like for example Telegram.
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:3]:
```suggestion
Bots are currently only possible in IRC style. They are entirely text based and not interactive like, for example, Telegram bots.
```
Telegram is interactive? or the bots?

File: proposals/3006-bot-interactions.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        "description": "The Search command",
+        "name": "Search"
+      }
+    ]
+  },
+  "type": "m.interaction.stage.start"
+}
+```
+
+6. The both now directly shows the new custom keyboard indicating the user currently is in a stage of the previous command.
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:80]:
```suggestion
6. The bot now directly shows the new custom keyboard, indicating that the user currently is in a stage of the previous command.
```
Some more grammar nitpicking here.

File: proposals/3006-bot-interactions.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+This is highly inspired by https://core.telegram.org/bots#keyboards and is trying to allow similar types 
+of keyboards for bots to allow more "fancy" bots while keeping full backwards compatibility.
+
+## Proposal
+
+The proposal is to have multiple interaction events. Which tell the client what kind of interactions the 
+bot has and the client replies with a specific format on that.
+
+_Note a deeper explanation of the events will be below the interaction flow_
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:14]:
```suggestion
_Note: a deeper explanation of the events will be below the interaction flow_
```
Not absolutely necessary, but...

File: proposals/3006-bot-interactions.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    }
+  ]
+}
+```
+
+#### Interaction event types
+
+Both `m.interaction.command` and `m.interaction.button` are required to be an object consisting of:
+
+* It's `type`
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:186]:
```suggestion
* Its `type`
```

File: proposals/3006-bot-interactions.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+This event indicates a flow end.
+
+The `event_id` is supposed to reference the event that was the start of the last event. It is not supposed to be the initial command.
+
+This event can be sent by both the client as well as the bot. 
+If a client sends the event the bot should think of this as if the user aborted the current action entirely.
+If a bot sends the event the client should think of this as having completed a flow fully.
+
+This event should be ignored when the user not started a command or the user has not previously received a stage start event.
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:255]:
```suggestion
This event should be ignored when the user has not started a command or the user has not previously received a stage start event. 
```

File: proposals/3006-bot-interactions.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+  "content": {
+    "event_id": "$random_id123"
+  },
+  "type": "m.interaction.stage.end"
+}
+```
+
+## What is `m.interaction.button`?
+
+This type is meant to be like telegrams custom keyboard feature. It is used to display a button in the UI which behaves like a custom keyboard with fixed functions.
```
Comment (LorenDB) [inline proposals/3006-bot-interactions.md:276]:
```suggestion
This type is meant to be like Telegram's custom keyboard feature. It is used to display a button in the UI which behaves like a custom keyboard with fixed functions.
```

**Review by LorenDB — COMMENTED** [review]
Just some grammar nitpicking. (Sorry!)

File: proposals/3006-bot-interactions.md (hunk 1 above)
Comment (MTRNord) [reply proposals/3006-bot-interactions.md:3]: Telegram bots were meant here :)

**Comment by penn5** [comment]
Why does there have to be a command prefix? Why does it have to be the same for all commands?

File: proposals/3006-bot-interactions.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+### Interaction Flow
+
+1. The bot joins a room and sends a state event containing information about possible interactions:
+
+```json
+{
+  "type": "m.bot.interactions",
+  "state_key": "@faq_bot:example.matrix",
+  "command_prefix": "!",
```
Comment (MTRNord) [inline proposals/3006-bot-interactions.md:24]: From @penn5 :
> Why does there have to be a command prefix? Why does it have to be the same for all commands?

The reason for the command prefix is to have it backwards compatible for cunstruction of the legacy message. From all bots I have seen so far in the matrix network this prefix always was the same across all commands the bot had. That is why I made this a extra key instead of having that in every interaction. Also doing it in the interaction would have added an additional "command" key to each interaction to allow for backwards compatibility.

Comment (penn5) [reply proposals/3006-bot-interactions.md:24]: Surely the current way, the command text is required to be equal to the name? Doesn't that mean that for backwards compatibility the name can't be human readable (or any more than the bot command used to be)? Surely it makes more sense to add a `text` key to each interaction to store the message that should be sent? Then, the bot is able to customise the fallback message...

File: proposals/3006-bot-interactions.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+
+## Potential issues
+
+* Priority of stages is not defined if 2 commands happen at the same time
+* Multiple users using multiple bots in a group is currently not handled directly. The MSC only demands to ignore stages if the user not started the command itself.
+
+## Alternatives
+
+* MSC2929 - This seems to solve parts of this as well but doesn't seem to specify custom keyboards
```
[the remark this answers is not in the document]
Comment (MTRNord) [reply proposals/3006-bot-interactions.md:294]: This proposal actually intentionally DOESNT allow complex flows. This is the result of the discussion that was in #matrix-client-developers:matrix.org . This is because at that point widgets are the correct way of doing this. Anything more complex than this most likely should be a widget as for client devs it gets increasingly difficult to add interactions more complex than this that also still have a uniform experience across all clients. 

TL;DR: This MSC intentionally limits itself to a small subset of interactions due to a discussion with various active client developers.

File: proposals/3006-bot-interactions.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+Imagine stages as submenus. They are indicating the next step in a greater action. 
+
+If a stage start is sent the user moves into the next decision tree of a menu. If a stage end happens the user gets back to the top of this decision tree.
+Stage end should not be treated as a back button.
+
+
+## Potential issues
+
+* Priority of stages is not defined if 2 commands happen at the same time
```
[the remark this answers is not in the document]
Comment (MTRNord) [reply proposals/3006-bot-interactions.md:289]: Timestamps are difficult across federated systems. However this most likely is already solved by the relations. (Would love to have some more input on this assumption from some that understand relations better than I do)

File: proposals/3006-bot-interactions.md (hunk 7 above)
Comment (MTRNord) [reply proposals/3006-bot-interactions.md:24]: Hm yeah that also makes sense.

File: proposals/3006-bot-interactions.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (erkinalp) [reply proposals/3006-bot-interactions.md:289]: > Timestamps are difficult across federated systems.

Read receipts only advance forwards, hence that order is always well defined and the problem boils down to retrieving the read receipts from the bot's server. And relations would not be useful across multiple unrelated commands from different users.

File: proposals/3006-bot-interactions.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+{
+  "content": {
+    "event_id": "$random_id123"
+  },
+  "type": "m.interaction.stage.end"
+}
+```
+
+7b. If the user clicks the new button the client sends the command as in step 4 with additionally mentioning the `stage_id` under the 
+`m.inteaction.command` object. The bot after that can react either with a new stage or with the same `m.interaction.stage.end`
```
[the remark this answers is not in the document]
Comment (erkinalp) [reply proposals/3006-bot-interactions.md:95]:
```suggestion
`m.interaction.command` object. The bot after that can react either with a new stage or with the same `m.interaction.stage.end` 
```

**Review by erkinalp — CHANGES_REQUESTED** [review]

**Review by erkinalp — APPROVED** [review]

**Comment by JoshuaDietz** [comment]
I like the interactions. However I see some problems with the buttons and the stages (if I understood the proposal right): As far as I understood a bot is a totally normal user account in matrix. This means that everyone can act like a bot in a room (i.e. send the events proposed here) and every room can have multiple members. Since it's namespaced with the username I don't see a problem with the interactions. But for the buttons there is the opportunity to do all kinds of strange stuff. I.e. in a room with multiple members a bot could start a stage on  user request. Then a completely other user can also send buttons. Now the client does not know which buttons to display. Even if this is circumvented, i.e. by the clients/server checking for which member/bot a command was intended and then allowing only this bot to show buttons is a problem because of the group nature of every room. I.e. imagine the following scenario:
You have a room with member A and B and bot C and D (all just users from a matrix standpoint).
1. Member A sends a command for bot C. 
2. Bot C presents it's buttons (to all members of the room I guess?) and the user waits a few seconds to decide which button to click.
3. Meanwhile member B sends a command to bot D.
4. Bot D presents it's buttons.

Now member A does not see the buttons of bot C with which he interacted but instead the ones of bot D (or possibly both depending on implementation)

Maybe I got this completely wrong and you planned a way to start something like a session when A sends the first command to bot C and then only the client of A displays the buttons and ignores other buttons or something like this. But if not I see a few opportunities for this to confuse users.

Maybe an easier approach could be the following:
1. Allow `<button>` tags in messages with a `button-data` attribute.
2. When a button is clicked the user who sent the message containing the buttons (i.e. the bot) gets a message notifying him that the button with the value of `button-data` was clicked by the user with id xy.
3. To present the next stage the bot can simply edit the message to display other buttons or remove them altogether.
4. 
The advantage of this approach is that the buttons are always bound to a message and only the sender of the message can edit them. That way presents much less opportunities to yield to confusion of the user because it's unclear to which bot the buttons belong to.

**Comment by erkinalp** [comment]
>     Allow <button> tags in messages with a button-data attribute.
>     When a button is clicked the user who sent the message containing the buttons (i.e. the bot) gets a message notifying him that the button with the value of button-data was clicked by the user with id xy.
>    To present the next stage the bot can simply edit the message to display other buttons or remove them altogether.

That has already been discussed in [matrix-spec room](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$a_gEoUFfq8Up_hc3HVMwR73P3hoES-3_UbJaAHtnbxg?via=t2l.io&via=matrix.org&via=jboi.nl) and there have been strong objections, such as:

> Imo, and I think some portion of the SCT shares this (judging by the matrix live episodes I’ve watched yesterday), HTML is turning out to be not that ideal for a message format, and tulir’s JSON-based formatting MSC was mentioned multiple times

**Comment by penn5** [comment]
> > Allow `<button>` tags in messages with a `button-data` attribute.
> > When a button is clicked the user who sent the message containing the buttons (i.e. the bot) gets a message notifying him that the button with the value of `button-data` was clicked by the user with id `xy`.
> > To present the next stage the bot can simply edit the message to display other buttons or remove them altogether.
> 
> That has already been discussed in [matrix-spec room](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$a_gEoUFfq8Up_hc3HVMwR73P3hoES-3_UbJaAHtnbxg?via=t2l.io&via=matrix.org&via=jboi.nl) and there have been strong objections, such as:
> 
> > Imo, and I think some portion of the SCT shares this (judging by the matrix live episodes I’ve watched yesterday), HTML is turning out to be not that ideal for a message format, and tulir’s JSON-based formatting MSC was mentioned multiple times

Then a Telegram-style approach, with buttons a part of the message in their own 2D array, could be more flexible. It would also remove the confusion that could occur, e.g., if a button appeared in the middle of a link

**Comment by JoshuaDietz** [comment]
> > Allow `<button>` tags in messages with a `button-data` attribute.
> > When a button is clicked the user who sent the message containing the buttons (i.e. the bot) gets a message notifying him that the button with the value of `button-data` was clicked by the user with id `xy`.
> > To present the next stage the bot can simply edit the message to display other buttons or remove them altogether.
> 
> That has already been discussed in [matrix-spec room](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$a_gEoUFfq8Up_hc3HVMwR73P3hoES-3_UbJaAHtnbxg?via=t2l.io&via=matrix.org&via=jboi.nl) and there have been strong objections, such as:
> 
> > Imo, and I think some portion of the SCT shares this (judging by the matrix live episodes I’ve watched yesterday), HTML is turning out to be not that ideal for a message format, and tulir’s JSON-based formatting MSC was mentioned multiple times

Thank you for the link. My main point was that buttons should be bound to a message. I agree that HTML is not ideal. I meant this to be more of a minimal example. The approach that every message can get a number of buttons in a special format sounds good to me.

File: proposals/3006-bot-interactions.md [hunk 11]
```diff
@@ -0,0 +1,322 @@
+# MSC3006: Bot Interactions
```
Comment (MTRNord) [inline proposals/3006-bot-interactions.md:1]: > Thank you for the link. My main point was that buttons should be bound to a message. I agree that HTML is not ideal. I meant this to be more of a minimal example. The approach that every message can get a number of buttons in a special format sounds good to me.

Moving from https://github.com/matrix-org/matrix-doc/pull/3006#issuecomment-830638719 (cc @JoshuaDietz ) to a thread. (Next time please use thread to make the PR more readable please)

The goal of this MSC is actually to not do inline buttons. That would be a separate MSC and I do agree it should exist. This one is meant to be non message bound commands :) Thats why buttons are not bound to messages.

Comment (JoshuaDietz) [reply proposals/3006-bot-interactions.md:1]: Yes I got that the goal of this MSC are not inline buttons. But as I described I don't really see how these keyboard buttons can work with matrix since every user can send them to every room as he wants (as long as there are no protections against that, i.e. a new permission). So if they are implemented in my opinion there needs to be a concept on how to avoid abuse of the feature (see the problems that I described in my previous comment)

Comment (MTRNord) [reply proposals/3006-bot-interactions.md:1]: Well a) as "m.bot.interactions" is an own event you could restrict it by powerlevels b) thats more an overall "issue" in matrix if you want to call it like that. Because bots are not separate from users. They are the same thing. So solving this is imho out of scope of this MSC. This is only possible to solve by separating users and bots as otherwise it is impossible to make restrictions that work for bots but restrict users :/

**Comment by MTRNord** [comment]
Dropping this for personal reasons

**Comment by MTRNord** [comment]
Reopening as I have some more capacity again for this
